File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/SupportsPushDownCatalystFilters.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.spark.sql.sources.v2.reader;

import org.apache.spark.annotation.InterfaceStability;
import org.apache.spark.sql.catalyst.expressions.Expression;

/**
 * A mix-in interface for {@link DataSourceReader}. Data source readers can implement this
 * interface to push down arbitrary expressions as predicates to the data source.
 * This is an experimental and unstable interface as {@link Expression} is not public and may get
 * changed in the future Spark versions.
 *
 * Note that, if data source readers implement both this interface and
 * {@link SupportsPushDownFilters}, Spark will ignore {@link SupportsPushDownFilters} and only
 * process this interface.
 */
@InterfaceStability.Unstable
public interface SupportsPushDownCatalystFilters extends DataSourceReader {

  /**
   * Pushes down filters, and returns filters that need to be evaluated after scanning.
   */
  Expression[] pushCatalystFilters(Expression[] filters);

  /**
   * Returns the catalyst filters that are pushed to the data source via
   * {@link #pushCatalystFilters(Expression[])}.
   *
   * There are 3 kinds of filters:
   *  1. pushable filters which don't need to be evaluated again after scanning.
   *  2. pushable filters which still need to be evaluated after scanning, e.g. parquet
   *     row group filter.
   *  3. non-pushable filters.
   * Both case 1 and 2 should be considered as pushed filters and should be returned by this method.
   *
   * It's possible that there is no filters in the query and
   * {@link #pushCatalystFilters(Expression[])} is never called, empty array should be returned for
   * this case.
   */
  Expression[] pushedCatalystFilters();
}

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/connector/SupportsPushDownCatalystFilters.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.spark.sql.internal.connector

import org.apache.spark.sql.catalyst.expressions.Expression
import org.apache.spark.sql.connector.expressions.filter.Predicate
import org.apache.spark.sql.connector.read.ScanBuilder

/**
 * A mix-in interface for {@link ScanBuilder}. Data sources can implement this interface to
 * push down filters to the data source. The pushed down filters will be separated into partition
 * filters and data filters. Partition filters are used for partition pruning and data filters are
 * used to reduce the size of the data to be read.
 */
trait SupportsPushDownCatalystFilters extends ScanBuilder {

  /**
   * Pushes down catalyst Expression filters (which will be separated into partition filters and
   * data filters), and returns data filters that need to be evaluated after scanning.
   */
  def pushFilters(filters: Seq[Expression]): Seq[Expression]

  /**
   * Returns additional filters inferred from eligible query filters passed to [[pushFilters]].
   * Each inferred filter must be implied by those query filters and satisfied by every row
   * returned by the scan.
   *
   * When `SupportsReportStatistics.reflectsFullyPushedDownFilters` returns `false`, Spark adds
   * inferred filters whose columns remain in the scan output to the logical Filter for optimizer
   * statistics. As with fully pushed filters, columns are not retained solely for this adjustment,
   * and filters that reference pruned columns are dropped.
   *
   * Spark discards inferred filters if a join, aggregate, or variant extraction replaces the scan
   * output.
   *
   * Inferred filters must be deterministic, contain no subqueries, user-defined expressions,
   * aggregate expressions, window expressions, or generators, resolve to well-typed Boolean
   * expressions, and not duplicate fully pushed filters. Spark ignores invalid inferred filters.
   *
   * Column references must be represented by `AttributeReference`. A nested column is represented
   * by a dotted name, with path parts containing dots quoted using Spark SQL identifier syntax.
   * For example, nested column `tz` in `location` is `location.tz`, while nested column `c.d` in
   * top-level column `a.b` is represented as `` `a.b`.`c.d` ``.
   */
  def inferredFilters: Seq[Expression] = Nil

  /**
   * Returns the data filters that are pushed to the data source via
   * {@link #pushFilters(Seq[Expression])}.
   */
  def pushedFilters: Array[Predicate]
}
